File: Python/Data Structures & Algorithms Module - Python/Week7/Palindrome.py
"""
Palindrome.py
Billy Cussen
18/11/2020
"""
import timeit

def palindrome(word):
    wordLength = len(word)
    match = True
    tempLen = wordLength-1
    for i in range(int(wordLength/2)):
        if word[i] != word[tempLen]:
            match = False
            break
        else:
            tempLen-=1
    return match

def palindromeRecursion(word):
    if len(word) == 0:
        return True
    
    if word[0] == word[len(word)-1]:
        return palindromeRecursion(word[1:len(word)-1])
    else:
        return False

startTime = timeit.default_timer()
print("Palindrome: "+str(palindrome("modom")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Palindrome Recursion: "+str(palindromeRecursion("moddom")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Palindrome: "+str(palindrome("hello")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Palindrome Recursion: "+str(palindromeRecursion("hello")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Palindrome: "+str(palindrome("racecar")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Palindrome Recursion: "+str(palindromeRecursion("racecar")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")

File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week7/Palindrome.java
/*
Palindrome.java
Billy Cussen
18/11/2020
*/
public class Palindrome{

    public static void main(String[]args){

        long startTime = System.nanoTime();
        System.out.println("Palindrome: "+palindrome("racecar"));
        long endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");
        
        startTime = System.nanoTime();
        System.out.println("Palindrome Recursion: "+palindromeRecursion("racecar"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Palindrome: "+palindrome("hello"));
        endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Palindrome Recursion: "+palindromeRecursion("hello"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");
    }

    public static boolean palindrome(String word){
        int lenTemp = word.length()-1;
        boolean pal = true;
        for(int i = 0; i < word.length()/2;i++){
            if(word.charAt(i)==word.charAt(lenTemp)){
                lenTemp--;
            } else{
                pal = false;
                break;
            }
        }
        return pal;
    }

    public static boolean palindromeRecursion(String word){
        if(word.length()<=1){
            return true;
        } 
        
        if(word.charAt(0)==word.charAt(word.length()-1)){
            return palindromeRecursion(word.substring(1,word.length()-1));
        } else{
            return false;
        }
    }
    

}